### ISSUE TITLE
Error: Bundles must have unique names.

### ISSUE THREAD

**Issue by gxy5202** [gxy5202/parcel-namer-hashless#1]
Error: Bundles must have unique names. The css file is referenced while using the style tag, they will be output as index.css at the same time

**Comment by pixu1980** [comment]
Same here

**Comment by pixu1980** [comment]
@gxy5202 nothing particular, static website building with just the specified ".parcelrc" in your readme.

`
  "dependencies": {
    "highlight.js": "^11.5.1",
    "reveal.js": "^4.3.1"
  },
  "devDependencies": {
    "@parcel/core": "^2.5.0",
    "@parcel/optimizer-svgo": "^2.5.0",
    "@parcel/transformer-image": "^2.5.0",
    "@parcel/transformer-sass": "^2.5.0",
    "@parcel/workers": "^2.5.0",
    "parcel": "^2.5.0",
    "parcel-namer-hashless": "^1.0.3",
    "posthtml-doctype": "^1.1.1",
    "posthtml-include": "^1.7.3"
  },
`

**Comment by gxy5202** [comment]
@pixu1980 yeah, llike which imported files are packaged in your project, or you can send a screenshot of the error report

**Comment by pixu1980** [comment]
@gxy5202 it fails while building font files assets i guess

**Comment by gxy5202** [comment]
@pixu1980 Thanks. It seems that the mechanism of parcel, the name of the compiled static file must be a unique value with a hash. But in dev mode, parcel-namer-hashless won't work.Does this happen in your production mode?

**Comment by gxy5202** [comment]
@pixu1980 I noticed that you used NODE_ENV=development, remove this command and try again

**Comment by pixu1980** [comment]
@gxy5202 same result

**Comment by gxy5202** [comment]
@pixu1980 Thanks for your feedback, I'm going to check the parcel source code, and see if there is any solution to solve this problem. Before that, you may use `parcel-namer-rewrite` plugin to change your bundle file name.

**Comment by gxy5202** [comment]
@pixu1980 Thanks for your feedback, I'm going to check the parcel source code, and see if there is any solution to solve this problem. Before that, you may use `parcel-namer-rewrite` plugin to change your bundle file name.

**Comment by devotion-sam** [comment]
This happens when parcel does code splitting. It seems to name the split chunk the same as whatever first referenced it.
For example if module1.js and module2.js both import jquery, instead of building something like

module1.hash.js
module2.hash.js
jquery.hash.js

You will end up with

module1.hash1.js
module1.hash2.js (this one is jquery)
module2.hash.js

I'm not sure where the responsibility for this lies but I don't think it's in this namer plugin. It does what it's meant to do which is remove the hashes from filenames. Perhaps this problem is fixable by using a custom bundler plugin? I have no idea.

Another thing I'm not sure where the responsibility lies is that when I use parcel to build an HTML file that has links to scripts and styles in it, the files will have the hashes removed, but the links to those files in the built HTML will not.

**Comment by gxy5202** [comment]
> seems
Thanks for your feedback.
1.It looks like you're describing the 'Shared Boundles' problem.The only thing I'm sure of is that the error was reported inside parcel, so I don't know how to solve this problem right now.
2.Can you provide a more detailed description of the built HTML issue？

**Comment by devotion-sam** [comment]
1. To get around this I'm disabling code splitting with

```
"@parcel/bundler-default": {
    "minBundles": 100,
    "minBundleSize": 3000,
    "maxParallelRequests": 20
  }
```

2. Apologies - after deleting .parcel-cache this is no longer an issue

**Comment by RichardJECooke** [comment]
I'm having the same error with `.js` files not `css` files. Even though I have only one `.js` bundle output at the end of Parcel, I have identical filenames in various folders and I think that is breaking this naming plugin. Any idea what to do please? I supposed could check every filename for duplicates and change them to be unique...

**Comment by gxy5202** [comment]
> I'm having the same error with `.js` files not `css` files. Even though I have only one `.js` bundle output at the end of Parcel, I have identical filenames in various folders and I think that is breaking this naming plugin. Any idea what to do please? I supposed could check every filename for duplicates and change them to be unique...

Which version are you useing? (parcel and parcel-namer-hashless)

**Comment by RichardJECooke** [comment]
Parcel 2.8.3. And this plugin 1.0.5

**Comment by gxy5202** [comment]
> Parcel 2.8.3. And this plugin 1.0.5


OK, could you please list your file structure(just fake is ok) or give me an example(online code tools like codepen) that I can test it, then i think i can give you a solution.

**Comment by RichardJECooke** [comment]
Huh, I actually don't have any unique filenames, that's weird:

```bash
./src/serviceWorker.ts

./src/js/2_logic/state.ts
./src/js/2_logic/helper.ts

./src/js/3_io/browserStore.ts
./src/js/3_io/dropboxStore.ts
./src/js/3_io/storeHelper.ts

./src/js/1_data/types.ts
./src/js/1_data/testData.ts
./src/js/1_data/messages.ts
./src/js/1_data/constants.ts

./src/js/4_widgets/app.vue

./src/js/4_widgets/io/dropboxSignout.vue
./src/js/4_widgets/io/dropboxOAuthRedirectPage.vue
./src/js/4_widgets/io/save.vue
./src/js/4_widgets/io/dropboxSignin.vue

./src/js/4_widgets/home/home.vue

./src/js/4_widgets/part/header.vue

./src/js/4_widgets/journal/journalScreen.vue
./src/js/4_widgets/journal/text.vue
./src/js/4_widgets/journal/date.vue
./src/js/4_widgets/journal/time.vue
./src/js/4_widgets/journal/journalEntry.vue
./src/js/4_widgets/journal/deleteEntryButton.vue
./src/js/4_widgets/journal/mood.vue
./src/js/4_widgets/journal/newEntryButton.vue
./src/js/4_widgets/router.ts

./src/index.html

./src/external/css/normalize.css

./src/media/img
./src/media/img/logo.svg
./src/media/img/appIcon512.png
./src/media/img/logo.png

./src/main.ts
./src/app.webmanifest

./src/css
./src/css/main.css
```

I found this command works fine to remove content hashes instead of your name plugin though:

```bash
#!/bin/bash

for file in ./dist/*.*; do
    new_filename=$(basename "$file" | sed -E 's/\.[a-zA-Z0-9]+\.(\w+)$/\.\1/')
    if [[ "$new_filename" != "$(basename "$file")" ]]; then
        mv "$file" "./dist/$new_filename"
    fi
done
```

**Comment by RichardJECooke** [comment]
I even tried specifying only the final output `index.js`, but it doesn't work: 

```
"include": ["serviceWorker.js$", "index.html$", "css$", "png$", "svg$", "index.js$"],
```

**Comment by gxy5202** [comment]
> Huh, I actually don't have any identical filenames, that's weird:
> 
> ```shell
> ./src/serviceWorker.ts
> 
> ./src/js/2_logic/state.ts
> ./src/js/2_logic/helper.ts
> 
> ./src/js/3_io/browserStore.ts
> ./src/js/3_io/dropboxStore.ts
> ./src/js/3_io/storeHelper.ts
> 
> ./src/js/1_data/types.ts
> ./src/js/1_data/testData.ts
> ./src/js/1_data/messages.ts
> ./src/js/1_data/constants.ts
> 
> ./src/js/4_widgets/app.vue
> 
> ./src/js/4_widgets/io/dropboxSignout.vue
> ./src/js/4_widgets/io/dropboxOAuthRedirectPage.vue
> ./src/js/4_widgets/io/save.vue
> ./src/js/4_widgets/io/dropboxSignin.vue
> 
> ./src/js/4_widgets/home/home.vue
> 
> ./src/js/4_widgets/part/header.vue
> 
> ./src/js/4_widgets/journal/journalScreen.vue
> ./src/js/4_widgets/journal/text.vue
> ./src/js/4_widgets/journal/date.vue
> ./src/js/4_widgets/journal/time.vue
> ./src/js/4_widgets/journal/journalEntry.vue
> ./src/js/4_widgets/journal/deleteEntryButton.vue
> ./src/js/4_widgets/journal/mood.vue
> ./src/js/4_widgets/journal/newEntryButton.vue
> ./src/js/4_widgets/router.ts
> 
> ./src/index.html
> 
> ./src/external/css/normalize.css
> 
> ./src/media/img
> ./src/media/img/logo.svg
> ./src/media/img/appIcon512.png
> ./src/media/img/logo.png
> 
> ./src/main.ts
> ./src/app.webmanifest
> 
> ./src/css/main.css
> ```
> 
> I found this command works fine to remove content hashes instead of your name plugin though:
> 
> ```shell
> #!/bin/bash
> 
> for file in ./dist/*.*; do
>     new_filename=$(basename "$file" | sed -E 's/\.[a-zA-Z0-9]+\.(\w+)$/\.\1/')
>     if [[ "$new_filename" != "$(basename "$file")" ]]; then
>         mv "$file" "./dist/$new_filename"
>     fi
> done
> ```

Yes, technically u can use any tools or language to change a file‘s name if they have the ability.  I think your problem is because of a <script> or a <style> tag in html? Maybe set includes only .js$ at first then try again. I will use vue to try this, see if i can solve this problem.

**Comment by gxy5202** [comment]
> > "@parcel/bundler-default": {
> > "minBundles": 100,
> > "minBundleSize": 3000,
> > "maxParallelRequests": 20
> > }
> 
> @devotion-sam , what does this do please? Surely if you want to disable splitting then the max of something should be 1 somewhere...
> 
> And where do I put that code??

see this doc. In package.json
https://parceljs.org/features/code-splitting/

**Comment by RichardJECooke** [comment]
> > Parcel 2.8.3. And this plugin 1.0.5
> 
> OK, could you please list your file structure(just fake is ok) or give me an example(online code tools like codepen) that I can test it, then i think i can give you a solution.

The bundles config doesn't help either. Here's the repository, with keys removed - https://we.tl/t-rdGd2WJe0E.

If you have time to see why your plugin isn't working I'd appreciate it.

To run:
```bash
npm update
npm run buildNoCheck
```

**Comment by gxy5202** [comment]
> > > Parcel 2.8.3. And this plugin 1.0.5
> > 
> > OK, could you please list your file structure(just fake is ok) or give me an example(online code tools like codepen) that I can test it, then i think i can give you a solution.
> 
> The bundles config doesn't help either. Here's the repository, with keys removed - https://we.tl/t-rdGd2WJe0E.
> 
> If you have time to see why your plugin isn't working I'd appreciate it.
> 
> To run:
> ```bash
> npm update
> npm run buildNoCheck
> ```

OK, I will check it later and give you a response

**Comment by gxy5202** [comment]
> > > Parcel 2.8.3. And this plugin 1.0.5
> > 
> > 
> > OK, could you please list your file structure(just fake is ok) or give me an example(online code tools like codepen) that I can test it, then i think i can give you a solution.
> 
> The bundles config doesn't help either. Here's the repository, with keys removed - https://we.tl/t-rdGd2WJe0E.
> 
> If you have time to see why your plugin isn't working I'd appreciate it.
> 
> To run:
> 
> ```shell
> npm update
> npm run buildNoCheck
> ```

Sorry to keep you waiting, I finally found the problem.

There is a <script> tag in your index.html file. In this case, parcel will generate a index.js file. So if you have another <script> tag to import 'main.ts', parcel will generate two index.js file. But we should notice that the first index.js file will not output as a file in your dist. So i delete your first <script> tag, then everything works fine.

![image](https://github.com/gxy5202/parcel-namer-hashless/assets/37973863/dc030890-0109-4f7e-a531-c1bc82cb159f)

**Comment by gxy5202** [comment]
Fixed (#10 ). Please update to v1.0.6.
